fedscale_examples/Async_FL/async_aggregator.py

```diff
@@ -0,0 +1,489 @@
+        but each key corresponds to a group of weights (e.g., for Tensorflow)"""
+
+        for p_g in results['update_weight']:
+            param_weights = results['update_weight'][p_g]
+            for idx, param_weight in enumerate(param_weights):
+                if isinstance(param_weight, list):
+                    param_weight = np.asarray(param_weight, dtype=np.float32)
+                param_weight = torch.from_numpy(param_weight).to(device=self.device)
+
+                if self.model_in_update == 1:
+                    self.model_weights[p_g][idx].data = param_weight
+                else:
+                    self.model_weights[p_g][idx].data += param_weight
+
+        if self.model_in_update == self.tasks_round:
+            for p in self.model_weights:
+                for idx in range(len(self.model_weights[p])):
+                    d_type = self.model_weights[p][idx].data.dtype
+
+                    self.model_weights[p][idx].data = (
+                            self.model_weights[p][idx].data / float(self.tasks_round)
+                    ).to(dtype=d_type)
+
+    def round_weight_handler(self, last_model):
+        """Update model when the round completes"""
+        if self.round > 1:
+            if self.args.engine == events.TENSORFLOW:
+                for layer in self.model.layers:
+                    layer.set_weights([p.cpu().detach().numpy() for p in self.model_weights[layer.name]])
+
+            else:
+                self.model.load_state_dict(self.model_weights)
+                current_grad_weights = [param.data.clone() for param in self.model.parameters()]
+                self.optimizer.update_round_gradient(last_model, current_grad_weights, self.model)
+
+    def round_completion_handler(self):
+        self.global_virtual_clock = self.round_stamp[-1] # += self.round_duration
+        # self.round_stamp.append(self.global_virtual_clock)
+        self.round += 1
+
+        if self.round % self.args.decay_round == 0:
+            self.args.learning_rate = max(self.args.learning_rate * self.args.decay_factor, self.args.min_learning_rate)
+
+        # handle the global update w/ current and last
+        self.round_weight_handler(self.last_gradient_weights)
+
+        avg_loss = sum(self.loss_accumulator) / max(1, len(self.loss_accumulator))
+        logging.info(f"Wall clock: {round(self.global_virtual_clock)} s, round: {self.round}, Remaining participants: " + \
+                     f"{self.resource_manager.get_remaining()}, Succeed participants: " + \
+                     f"{len(self.stats_util_accumulator)}, Training loss: {avg_loss}")
+
+        # dump round completion information to tensorboard
+        if len(self.loss_accumulator):
+            self.log_train_result(avg_loss)
+
+        # update select participants
+        if self.round % 50 == 1: #  total_worker > buffer_size * 100
+            self.sampled_participants = self.select_participants(
```

Comment:
Why does the server samples participants every 50 rounds? For my understanding, 50 rounds are 50*`async_buffer_size`=500 clients' updates. But the server samples `total_worker=800` clients a time. The overcommitment seems too large. Or maybe a better idea is to have a parameter controlling the overcommitment.

Reply:
it just simulates the constant client arrival

Reply:
see the above reply

Reply:
1. do not use magic numbers. Please change 50 to a configuration. 2. Yes. we already have the configurable overcommitment.

Reply:
there is no `overcommitment` in Fedbuff async setting. They just assume constant arrival rate and assign the latest model version, and the job fails if the client goes offline